### PR TITLE
post workflow outputs back to entity model

### PR DESCRIPTION
**Pull request by dvoet**

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala [hunk 1]
```diff
@@ -208,6 +208,7 @@ trait GraphDAO {
           case v => Option(v)
         })
       }
+      x
```
Comment (abaumann) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: why this change?

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpExecutionServiceDAO.scala [hunk 2]
```diff
@@ -30,4 +30,11 @@ class HttpExecutionServiceDAO( executionServiceURL: String )( implicit system: A
     val pipeline = addHeader(Cookie(authCookie)) ~> sendReceive ~> unmarshal[ExecutionServiceStatus]
     Await.result(pipeline(Get(url)),Duration.Inf)
   }
+
+  override def outputs(id: String, authCookie: HttpCookie): ExecutionServiceOutputs = {
+    val url = executionServiceURL + s"/workflow/${id}/outputs"
+    import system.dispatcher
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpExecutionServiceDAO.scala]: ...?

Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpExecutionServiceDAO.scala]: Is this some stupid scala infix thing?

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala (hunk 1 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: debug code, will revert

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpExecutionServiceDAO.scala (hunk 2 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpExecutionServiceDAO.scala]: what are you talking about?

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala [hunk 3]
```diff
@@ -228,6 +229,7 @@ trait GraphDAO {
           val workspaceVertex = getSinglePipelineResult(workspacePipeline(graph, workspaceName.namespace, workspaceName.name)).get
           new GremlinPipeline(vertex).out(workflowEdge).linkIn(workflowEdge, workspaceVertex).iterate()
         case Some(x: WorkflowFailure) => serializeDomainObjects(vertex, seq.asInstanceOf[Seq[WorkflowFailure]], workflowFailureEdge, (wf: WorkflowFailure) => (wf.entityType, wf.entityName), graph, workspaceName)
+        case Some(a: Attribute) => serializeSeq(graph, workspaceName, vertex, key, seq.asInstanceOf[Seq[Attribute]])
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala]: I know it's not really part of this PR but I'd like to register my displeasure at the serialization of `Seq[]` here. This is not very polymorphic-collectiony. Whatever happened to `seq.map( _.serialize )` ?

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpExecutionServiceDAO.scala (hunk 2 above)
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpExecutionServiceDAO.scala]: There appears to be an import in the middle of a function.

Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpExecutionServiceDAO.scala]: Not where imports usually live.

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowMonitor.scala [hunk 4]
```diff
@@ -56,7 +71,32 @@ class WorkflowMonitor(parent: ActorRef, pollInterval: Duration, workflow: Workfl
...
+        }
+      }
+
+      if (attributes.forall(_.isSuccess)) {
+        (workflow.copy(status = completionStatus), Option(attributes.map(_.get).toMap))
+
+      } else {
+        val errors = attributes.collect { case Failure(t) => AttributeString(t.getMessage) }
+        (workflow.copy(messages = Option(errors.toSeq), status = WorkflowStatuses.Failed), None)
+      }
```
Comment (abaumann) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowMonitor.scala]: i like this pattern for trying and collecting errors

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowMonitor.scala [hunk 5]
```diff
@@ -56,7 +71,32 @@ class WorkflowMonitor(parent: ActorRef, pollInterval: Duration, workflow: Workfl
...
+      val attributes = methodConfig.outputs.map { case (outputName, attributeName) =>
+        Try {
+          attributeName.value -> outputs.getOrElse(outputName, {
+            throw new RawlsException(s"output named ${outputName} does not exist")
+          })
+        }
+      }
+
+      if (attributes.forall(_.isSuccess)) {
+        (workflow.copy(status = completionStatus), Option(attributes.map(_.get).toMap))
```
Comment (abaumann) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowMonitor.scala]: later if we resubmit due to some error (say output was not available at time and became available again), do we need to clear messages on workflow here?

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpExecutionServiceDAO.scala (hunk 2 above)
Comment (mattsooknah) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpExecutionServiceDAO.scala]: I've noticed that this particular import ( ```system.dispatcher``` ) consistently shows up in the middle of functions. I suspect it's some magic thing related to Akka/Spray.

Comment (geoffjentry) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpExecutionServiceDAO.scala]: If you're pulling in implicits (which obviously akka uses extensively) it's a way to make sure that the right stuff is being used in the right places. Otherwise I agree that imports should generally be at the top

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowMonitor.scala [hunk 6]
```diff
@@ -56,7 +71,32 @@ class WorkflowMonitor(parent: ActorRef, pollInterval: Duration, workflow: Workfl
     }
   }
 
-  def onWorkflowDone(updateWorkflow: Workflow, txn: RawlsTransaction): Unit = {
-    // call any code to update entity model here
+  def withMethodConfig(txn: RawlsTransaction)(op: MethodConfiguration => (Workflow, Option[Map[String, Attribute]])): (Workflow, Option[Map[String, Attribute]]) = {
```
Comment (mattsooknah) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowMonitor.scala]: Is there a reason why we can't replace all the instances of ```(Workflow, Option[Map[String, Attribute]])``` with ```WorkflowStatusChange``` ? It's clearer and less verbose.

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowMonitor.scala (hunk 5 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowMonitor.scala]: if we resubmit it should be a new workflow

**Comment by mattsooknah** [comment]
looks good to me :+1:
